src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -132,41 +177,42 @@ def _backchain_one_pass(
                 # They must be a subset of the current PNAD's add effects.
                 necessary_add_effects = necessary_image - atoms_seq[t]
                 assert necessary_add_effects.issubset(segment.add_effects)
+                # Update the segment's necessary_add_effects.
+                segment.necessary_add_effects = necessary_add_effects
 
                 # We start by checking if any of the PNADs associated with the
                 # demonstrated option are able to match this transition.
-                for pnad in pnads_for_option:
-                    ground_op = self._find_unification(necessary_add_effects,
-                                                       pnad, segment)
-                    if ground_op is not None:
-                        obj_to_var = dict(
-                            zip(ground_op.objects, pnad.op.parameters))
-                        if len(param_opt_to_nec_pnads[option.parent]) == 0:
-                            param_opt_to_nec_pnads[option.parent].append(pnad)
-                        break
+                pnad, ground_op = self._find_unification(
+                    necessary_add_effects, pnads_for_option, segment)
+                if pnad is not None:
+                    assert ground_op is not None
+                    obj_to_var = dict(
+                        zip(ground_op.objects, pnad.op.parameters))
+                    if len(param_opt_to_nec_pnads[option.parent]) == 0:
+                        param_opt_to_nec_pnads[option.parent].append(pnad)
                 # If we weren't able to find a substitution (i.e, the above
                 # for loop did not break), we need to try specializing each
                 # of our PNADs.
                 else:
                     nec_pnad_set_changed = True
-                    for pnad in pnads_for_option:
-                        new_pnad = self._try_specializing_pnad(
-                            necessary_add_effects, pnad, segment)
-                        if new_pnad is not None:
-                            assert new_pnad.option_spec == pnad.option_spec
-                            if len(param_opt_to_nec_pnads[option.parent]) > 0:
-                                param_opt_to_nec_pnads[option.parent].remove(
-                                    pnad)
-                            del pnad
-                            break
+                    specialization_output = self._try_specializing_pnad(
+                        necessary_add_effects, pnads_for_option, segment)
+                    if specialization_output is not None:
+                        new_pnad, pnad = specialization_output
```

Comment:
this seems like a hacky way to do what you want... 
can you do the first part of `try_specializing_pnad()` here instead? namely, the calling of `find_unification` to obtain the `pnad`. then, you can just pass that `pnad` into `try_specializing_pnad()` instead of passing in the whole `pnads_for_option`

Reply:
good point - i can definitely refactor to make it like this!